ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,188 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    /* Sanity check */
+    ASSERT(SymbolicLinkName != NULL);
+
+    /* Convert alias GUID to a string */
+    Status = RtlStringFromGUID(AliasInterfaceClassGuid, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    BaseKeyName.Length = 0;
+    BaseKeyName.MaximumLength += GuidString.Length;
+    BaseKeyName.Buffer = ExAllocatePoolWithTag(NonPagedPool,
+                                               BaseKeyName.MaximumLength,
+                                               TAG_IO);
+    if (!BaseKeyName.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    /* Copy alias GUID string to a base key */
+    RtlCopyUnicodeString(&BaseKeyName, &GuidString);
+
+    /* Open base registry key */
+    Status = IoOpenDeviceInterfaceRegistryKey(&BaseKeyName,
+                                              GENERIC_READ,
+                                              &InterfaceKey);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("IoOpenDeviceInterfaceRegistryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        NULL,
+                        0,
+                        &Size);
+    if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    FullInfo = ExAllocatePoolWithTag(PagedPool,
+                                     Size,
+                                     TAG_IO);
+    if (!FullInfo)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        FullInfo,
+                        Size,
+                        &Size);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Enumerate subkeys (i.e. the different device objects) */
+    while (TRUE)
+    {
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                NULL,
+                                0,
+                                &NeededLength);
+        if (Status == STATUS_NO_MORE_ENTRIES)
+        {
+            break;
+        }
+        else if (Status == STATUS_BUFFER_TOO_SMALL)
+        {
+            ActualLength = NeededLength;
+            if (BasicInfo) ExFreePoolWithTag(BasicInfo, TAG_IO);
+
+            BasicInfo = ExAllocatePoolWithTag(PagedPool,
```

Comment:
You're still unconditionally reallocating every iteration, since the first call with NULL/0 will always end up going through this path on success. But this does seem to fix the leak.